google-services-plugin/src/main/kotlin/com/google/gms/googleservices/GoogleServicesPlugin.kt

```diff
@@ -0,0 +1,153 @@
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import java.util.Locale
+
+class GoogleServicesPlugin : Plugin<Project> {
+
+    override fun apply(project: Project) {
+        val config =
+            project.extensions.create("googleServices", GoogleServicesPluginConfig::class.java)
+        project.afterEvaluate {
+            if (config.disableVersionCheck) {
+                return@afterEvaluate
+            }
+
+            val globalDependencies = DependencyAnalyzer()
+            val strictVersionDepInspector = DependencyInspector(
+                globalDependencies, project.name,
+                "This error message came from the google-services Gradle plugin, report" +
+                        " issues at https://github.com/google/play-services-plugins and disable by " +
+                        "adding \"googleServices { disableVersionCheck = true }\" to your build.gradle file."
+            )
+            project.configurations.all { projectConfig ->
+                if (projectConfig.name.contains("ompile")) {
+                    projectConfig.incoming.afterResolve(strictVersionDepInspector::afterResolve)
+                }
+            }
+        }
+
+        var pluginApplied = false
+
+        project.pluginManager.withPlugin("com.android.application") {
+            pluginApplied = true
+            project.extensions.configure(ApplicationAndroidComponentsExtension::class.java) {
+                it.registerSourceType(SOURCE_TYPE)
+                it.onVariants { variant ->
+                    handleVariant(variant, project)
+                }
+            }
+        }
+        project.pluginManager.withPlugin("com.android.dynamic-feature") {
+            pluginApplied = true
+            project.extensions.configure(DynamicFeatureAndroidComponentsExtension::class.java) {
+                it.registerSourceType(SOURCE_TYPE)
+                it.onVariants { variant ->
+                    handleVariant(variant, project)
+                }
+            }
+        }
+
+        project.afterEvaluate {
+            if (pluginApplied) {
+                return@afterEvaluate
+            }
+            project.logger.error(
+                "The google-services Gradle plugin needs to be applied on a project with " +
+                        "com.android.application or com.android.dynamic-feature."
+            )
+        }
+    }
+
+    private fun <T> handleVariant(
+        variant: T,
+        project: Project
+    ) where T : Variant, T : GeneratesApk {
+        val jsonToXmlTask = project.tasks.register(
+            "process${variant.name.capitalize()}GoogleServices",
+            GoogleServicesTask::class.java
+        ) {
+            it.missingGoogleServicesStrategy.set(
+                project.extensions.getByType(GoogleServicesPluginConfig::class.java)
+                    .missingGoogleServicesStrategy
+            )
+            it.googleServicesJsonFiles.set(variant.sources.getByName(SOURCE_TYPE).all)
```

Comment:
while I could back away from using the custom sources API (`sources.getByName(SOURCE_TYPE)`) and instead hardcode the paths as before (which would work fine), I cannot escape it in line 113:

```variant.sources.res?.addGeneratedSourceDirectory```

as that's the new way of contributing Android resources to the build, and the whole point of this PR is to remove usage of the old, deprecated AGP APIs and move to the new ones. 

How do you suggest we proceed with this, should we wait until the Sources interface is released as stable in AGP 8.x (probably not 8.0 though), or would it be enough if I talked to the AGP leads and got an OK from them that these APIs won't change and WILL become stable in 8.1 or 8.2?

Reply:
> if I talked to the AGP leads and got an OK from them that these APIs won't change and WILL become stable in 8.1 or 8.2?

This would be good enough to keep it- and would help ease most of this PR over :')